Mathlib/LinearAlgebra/Matrix/SVD/Reindex.lean

```diff
@@ -0,0 +1,107 @@
+of Fin n and the eigenvalues of the matrix Aᴴ⬝A, partitioned into
+- non-zero eigenvalues: (exactly A.rank) of them, see `fin_rank_equiv_eigs_conjTranspose_mul_self`
+- zero eigenvalues: (exactly n - A.rank) of them -/
+noncomputable def eigenColumnEquiv (A: Matrix (Fin M) (Fin N) 𝕂) :
+    (Fin N) ≃ (Fin A.rank) ⊕ (Fin (N - A.rank)) := by
+  let en := Equiv.sumCompl (fun i =>  (isHermitian_transpose_mul_self A).eigenvalues i ≠ 0)
+  let eₙᵣ : {i // ¬(isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} ≃ Fin (N - A.rank) :=
+    Fintype.equivFinOfCardEq (by rw [Fintype.card_subtype_compl, Fintype.card_fin,
+      rank_eq_card_pos_eigs_conjTranspose_mul_self])
+  exact Equiv.trans en.symm  (Equiv.sumCongr (finRankEquivEigsConjTransposeMulSelf A).symm eₙᵣ)
+
+/-- For matrix of size m × n and rank A.rank : we have an bijeciton between the elements
+of Fin (height) and the eigevalues of the matrix A⬝Aᴴ partitioned into
+- non-zero eigenvalues: (exactly A.rank) of them see `fin_rank_equiv_eigs_mul_conjTranspose`
+- zero eigenvalues: (exactly height - A.rank) of them -/
+noncomputable def eigenRowEquiv (A: Matrix (Fin M) (Fin N) 𝕂) :
+    (Fin M) ≃ (Fin A.rank) ⊕ (Fin (M - A.rank)) := by
+  let em := Equiv.sumCompl (fun i =>  (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0)
+  let eᵣ' : {i // (isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} ≃ Fin A.rank :=
+    Fintype.equivFinOfCardEq (by rw [rank_eq_card_pos_eigs_self_mul_conjTranspose])
+  let eₘᵣ : {i // ¬(isHermitian_mul_conjTranspose_self A).eigenvalues i ≠ 0} ≃ Fin (M - A.rank) :=
+    Fintype.equivFinOfCardEq (by rw [Fintype.card_subtype_compl, Fintype.card_fin,
+      rank_eq_card_pos_eigs_self_mul_conjTranspose])
+  exact Equiv.trans em.symm  (Equiv.sumCongr eᵣ' eₘᵣ)
+
+/-- When the eigenvalues of the matrix Aᴴ⬝A are partitioned using
+`equiv_fin_width_eigs_conjTranspose_mul_self` i.e. into non-zero and zero eigenvalues, any element
+from the second partition is obviously zero! -/
+@[simp]
+lemma eigen_eigenColumnEquiv_inr (A: Matrix (Fin M) (Fin N) 𝕂) (i: Fin (N - A.rank)) :
+  (isHermitian_transpose_mul_self A).eigenvalues
+    ((eigenColumnEquiv A).symm (Sum.inr i)) = 0 := by
+  unfold eigenColumnEquiv Equiv.sumCongr
+  simp only [ne_eq, Equiv.symm_trans_apply, Equiv.symm_symm, Equiv.coe_fn_symm_mk, Sum.elim_inr,
+    Equiv.sumCompl_apply_inr]
+  let eₙᵣ : {i // ¬(isHermitian_transpose_mul_self A).eigenvalues i ≠ 0} ≃ Fin (N - A.rank) := by
+    apply Fintype.equivFinOfCardEq
+    rw [Fintype.card_subtype_compl, Fintype.card_fin, rank_eq_card_pos_eigs_conjTranspose_mul_self]
+  exact Iff.mp Function.nmem_support ((eₙᵣ.symm i).prop)
+
+/-- When the eigenvalues of the matrix A⬝Aᴴ are partitioned using
+`equiv_fin_height_eigs_mul_conjTranspose` i.e. into non-zero and zero eigenvalues, any element from
+the second partition is obviously zero! -/
+@[simp]
+lemma eigen_eigenRowEquiv_inr (A: Matrix (Fin M) (Fin N) 𝕂) (i: Fin (M - A.rank)) :
```

Comment:
```suggestion
lemma eigen_eigenRowEquiv_inr (A : Matrix (Fin M) (Fin N) 𝕂) (i : Fin (M - A.rank)) :
```
and in all the lemmas / variables lines in this file

Reply:
Done!